### ISSUE TITLE
Recursive PL/pgSQL function for the Nth Fibonacci number is noticeably slower in YugabyteDB than in PostgreSQL

### ISSUE THREAD

**Issue by bllewell** [yugabyte/yugabyte-db#2596]
Using YugabyteDB Version 2.0.1.0 downloaded from the Internet. Running on my MacBook macOS Mojave Version 10.14.6. Cluster created with bare "yb-ctl create" — so RF=1, one node.

# Recursive testcase

I implemented the text-book recursive solution to calculate the Nth Fibonacci number. It's famously slower than a straightforward iterative solution. But it's a convenient way to burn a sizeable wall-clock time (i.e. long enough to allow reasonable measurement reproducibility) for the execution of a small program text so that performance can be compared between various environments. In particular, the time for this test is dominated by function call and return rather than straight arithmetic operations.
```
create function fib(n in integer)
  returns integer
  language 'plpgsql'
as $$
declare
  -- Use these constants to be 100% sure that there's no datatype conversion.
  zero constant integer not null := 0;
  one  constant integer not null := 1;
  two  constant integer not null := 2;
begin
  -- Define fib(n) this way.
  if n < one then
    return zero;
  elsif n = one then
    return one;
  else
    return fib(n - one) + fib(n - two);
  end if;
end
$$;
```
Notice that it does no SQL. I invoked it, using YugabyteDB (hereinafter YB) and PostgreSQL (hereinafter PG), thus:
```
\timing on
select fib(<some N>);
\timing off
```
The first invocation (for the 3rd Fibonacci number) served the purpose of compiling the function so that the time that this takes wouldn't interfere with the subsequent timings. It's tiny compared to the times for the 30th and the 35th Fibonacci number.

For completeness, I implemented, and timed, the same logic in Python:
```
import datetime

cnt = 0

def fib(n):
    global cnt
    cnt = cnt + 1
    if n < 1:
        return 0
    elif n == 1:
        return 1
    else:
        return (fib(n - 1) + fib(n - 2))

def time_it(n):
    t_start = datetime.datetime.now()
    f = fib(n)
    t = (datetime.datetime.now() - t_start).total_seconds()
    print(
        "\n\nfib(" + str(n) + "): "
        + str(f) + " | " + "cnt: " + str(cnt)
        + " took " + "{:.3f}".format(t) + " seconds"
    )

# --------------------------------------------------------------------------------

def main():
    global cnt

    cnt = 0
    time_it(3)

    cnt = 0
    time_it(30)

    cnt = 0
    time_it(35)

# --------------------------------------------------------------------------------

if __name__ == '__main__':
    main()
```
Notice that, unlike PG/plSQL, Python allows a function to access and modify a global variable. This feature allows the total number of recursive calls to `fib(n)` to be counted. (See the results, below. The number is huge.)

Again for completeness, I implemented the same logic in Oracle Database 18C (running on Linux in a Parallels VM on my MacBook) using PL/SQL:
```
create package pkg is
  function fib(n in integer) return integer;
  cnt integer not null := 0;
end pkg;
/
create package body pkg is
  function fib(n in integer) return integer
  is
    -- Use these constants to be 100% sure that there's no datatype conversion.
    zero constant integer not null := 0;
    one  constant integer not null := 1;
    two  constant integer not null := 2;
  begin
    cnt := cnt + 1;

    -- Define fib(n) this way.
    if n < one then
      return zero;
    elsif n = one then
      return one;
    else
      return fib(n - one) + fib(n - two);
    end if;
  end fib;
end pkg;
/
```
Notice that PL/SQL's support for packages again allows the use of a global variable to count the number of recursive calls. Here's the calling block (executed using SQL*Plus):
```
SET TIMING ON
declare
  fib_n integer not null := 0;
  n integer not null := 35;
begin
  pkg.cnt := 0;
  select pkg.fib(n) into fib_n from dual;
  print(pkg.cnt, n, fib_n);
end;
/
SET TIMING OFF
```

Here are the results (times in seconds):
```
 N    Answer    Nr recursive calls       YB      PG   ORCL   Python                 
--   --------   ------------------   ------   -----   ----   ------                  
 3          2                    5     1.90    0.95   0.05     0.00
30     832040              2692537    21.56    7.90   0.60     0.61
35    9227465             29860703   228.89   89.17   6.74     6.69
```
The answers for `fib(n)` for all four implementations agree—and they agree with the answers turned up by Internet search. And the recursive call counts from Python and ORCL agree with each other.

The times for Python and ORCL are similar to each other. PostgreSQL is an order of magnitude slower. And YugabyteDB is two to three times slower than PostgreSQL.

# Non-recursive testcase

I also did a test that removed the effect of function invocation and return and used CPU only for arithmetic operations, (looping, "if" tests—but no SQL). I implemented the Leibniz series for π (see [HERE](https://en.wikipedia.org/wiki/Leibniz_formula_for_π)):
```
π/4 = 1 - 1/3 + 1/5 -1/7 + 1/9 - ...
```
This formula converges very slowly, so it again provides a way to measure long execution times for a small source text. Here is the function:
```
create function leibniz_pi()
  returns numeric
  language 'plpgsql'
as $$
declare
  true_pi constant numeric(25, 15) not null := 3.141592654;
  one     constant numeric(25, 15) not null := 1.0;
  four    constant numeric(25, 15) not null := 4.0;
  tol     constant numeric(25, 15) not null := 5.0e-9;
  ub      constant numeric(25, 15) not null := one + tol;
  lb      constant numeric(25, 15) not null := one - tol;

  pi               numeric(25, 15) not null := one;
  d                numeric(25, 15) not null := 3.0;
  q                numeric(25, 15) not null := one;

  j                int              not null := 1;
  k                int              not null := 1;
  n                int              not null := 1;

  odd              boolean          not null := true;
begin
  loop
    pi := case odd
            when true then pi - 1/d
            else           pi + 1/d
          end;
    q := (pi*four)/true_pi;

    if j >= 5000000 then
      raise info 'k, q: % million, %',
        to_char(k*5, '999'),
        to_char(q, '0.99999999999');
      j := 0;
      k := k + 1;
    end if;

    d := d + 2;
    odd := not odd;
    j := j + 1;
    n := n + 1;

    exit when q between lb and ub;
  end loop;
  raise info '';
  raise info 'nr of iterations: %', to_char(n, '999,999,999');
  raise info '';
  return q;
end
$$;
```
I used the same datatype for all variables to avoid conversion concerns. I had to accommodate the large value that `d` grows to as the series converges. Both the YugabyteDB implementation and the PostgreSQL implementation report the same final estimate-to-true-π ratio (1.00000000500) and the same number of iterations (62,055,093) to get there. The timings for the two implementations are very similar, with YugabyteDB coming out slightly ahead.
```
   seconds
YB   160.3
PG   174.4
```
# Conclusion
It seems that the YugabyteDB implementation and the PostgreSQL implementation of PL/pgSQL perform very much the same as each other for arithmetic operations, looping, and "if" tests—but that YugabyteDB's implementation is noticeably slower than PostgreSQL's for function call and return.

**Comment by kmuthukk** [comment]
Did

`% sudo perf record -ag`

and then:

`% sudo perf report`

<img width="905" alt="Screen Shot 2019-10-15 at 2 15 59 PM" src="https://user-images.githubusercontent.com/13615376/66870488-69a29780-ef56-11e9-89ae-75911c4c3f85.png">

**Comment by kmuthukk** [comment]
cc: @d-uspenskiy , @m-iancu (in case you have ideas).

**Comment by kmuthukk** [comment]
I tried using `strace` on the postgres worker process to figure out what file the `load_external_function` stack in the perf report above was going after, and the lines were showing this:

```
stat("/home/centos/yugabyte-2.0.1.0/postgres/lib/plpgsql", 0x7ffff01b10d0) = -1 ENOENT (No such file or directory)
stat("/home/centos/yugabyte-2.0.1.0/postgres/lib/plpgsql.so", {st_mode=S_IFREG|0755, st_size=928664, ...}) = 0
stat("/home/centos/yugabyte-2.0.1.0/postgres/lib/plpgsql", 0x7ffff01b10d0) = -1 ENOENT (No such file or directory)
stat("/home/centos/yugabyte-2.0.1.0/postgres/lib/plpgsql.so", {st_mode=S_IFREG|0755, st_size=928664, ...}) = 0
stat("/home/centos/yugabyte-2.0.1.0/postgres/lib/plpgsql", 0x7ffff01b05c0) = -1 ENOENT (No such file or directory)
stat("/home/centos/yugabyte-2.0.1.0/postgres/lib/plpgsql.so", {st_mode=S_IFREG|0755, st_size=928664, ...}) = 0
```

**Comment by d-uspenskiy** [comment]
I've run tests with `fib(35)` on my PC. Results are:
PG time: 87.6 secs
YB debug time: 182.0 secs
YB release time: 151.7 secs

So YB is 1.73 times slower than vanilla postgres in this particular scenario

**Comment by m-iancu** [comment]
Fixed in 7fc66c8.
